tools/_setup_generation/step_knowledge_base.py

```diff
@@ -0,0 +1,198 @@
+# ##########################################################################################
+# Step to generate the list of knowledge base items in the knowledge base index page
+#
+# The knowledge base index page is generated from the tutorials, demos and tips folders.
+# Each folder contains a list of sub folders, each sub folder is a knowledge base item.
+# Each item has an index.md file with a header containing the necessary information to
+# generate the list of items in the knowledge base index page.
+#
+# The header of an item is a multiline header starting with '---' and ending with '---'.
+# The header contains the following information:
+# - title: The title of the item
+# - category: The category of the item (tutorials, demos or tips)
+# - type: The type of the item (code, video or article)
+# - data-keywords: A comma separated list of keywords
+# - short_description: A short description of the item
+#
+# ##########################################################################################
+import os
+import re
+from typing import List
+
+from .setup import SetupStep, Setup
+from .step_knowedge_base import FolderItem, FileItem, Item
+from .step_knowedge_base.exceptions import WrongHeader, NoHeader, NoIndexFile
+
+
+class KnowledgeBaseStep(SetupStep):
+
+    # The name of the knowledge base folder
+    KNOWLEDGE_BASE_FOLDER_NAME = "knowledge_base"
+
+    # The placeholders in the knowledge base index file to be replaced by the generated content
+    TUTORIALS_PLACEHOLDER = r"\[LIST_OF_TUTORIALS\]"
+    DEMOS_PLACEHOLDER = r"\[LIST_OF_DEMOS\]"
+    TIPS_PLACEHOLDER = r"\[LIST_OF_TIPS\]"
+
+    def __init__(self):
+        self.KB_INDEX_PATH = None
+        self.KB_INDEX_TEMPLATE_PATH = None
+        self.TUTORIALS_FOLDER_PATHS = None
+        self.TUTORIALS_INDEX_PATH = None
+        self.TUTORIALS_INDEX_TPL_PATH = None
+        self.DEMOS_FOLDER_PATHS = None
+        self.DEMOS_INDEX_PATH = None
+        self.DEMOS_INDEX_TPL_PATH = None
+        self.TIPS_FOLDER_PATHS = None
+        self.TIPS_INDEX_PATH = None
+        self.TIPS_INDEX_TPL_PATH = None
+        self.tutorials_on_kb = ""
+        self.tutorials = ""
+        self.demos_on_kb = ""
+        self.demos = ""
+        self.tips_on_kb = ""
+        self.tips = ""
+
+    def enter(self, setup: Setup):
+        self.KB_INDEX_PATH = os.path.join(setup.docs_dir, self.KNOWLEDGE_BASE_FOLDER_NAME, "index.md")
+        self.KB_INDEX_TEMPLATE_PATH = os.path.join(setup.docs_dir, self.KNOWLEDGE_BASE_FOLDER_NAME, "index.md_template")
```

Comment:
I typically do not create this second variable: 
<prefix>_TEMPLATE_PATH = <prefix>_PATH+"_template"

So you don't have to synchronize the two variables... 
But that works, obviously